### PR TITLE
[15.0][IMP] sale_planner_calendar: Design improves and new buttons in calendar planner event

### PR DESCRIPTION
**Pull request by pilarvargas-tecnativa**
Improvements made:

- Profile for calendar events that will allow an icon to be displayed in the kanban view of the visit. It can also be configured from a recurring event.
- Tree view to edit and add new profiles with their icons which will be fontawesome text strings (e.g fa-user).
- New buttons in the kanban view of the visit: Location, Whatsapp
- Some improvements in the structure to make it responsive with the new addition of buttons that may or may not be visible.

Before:

image
image

After:

image
image

cc @Tecnativa TT48632

@pedrobaeza @sergio-teruel @CarlosRoca13 please review

### CODE REVIEW

**Review by CarlosRoca13 — APPROVED** [review]
Tested on runboat :+1:

File: sale_planner_calendar/models/sale_planner_calendar_event.py [hunk 1]
```diff
@@ -130,6 +141,46 @@ def _compute_contact(self):
             )
             rec.partner_contact_name = contact.name
 
+    def _compute_sanitized_partner_mobile(self):
+        for rec in self:
+            rec.sanitized_partner_mobile = False
+            if rec.partner_mobile:
```
Comment (chienandalu) [inline sale_planner_calendar/models/sale_planner_calendar_event.py]:
```suggestion
        self.sanitized_partner_mobile = False
        for rec in self.filtered("partner_mobile"):
```

File: sale_planner_calendar/models/sale_planner_calendar_event.py [hunk 2]
```diff
@@ -73,6 +78,12 @@ class SalePlannerCalendarEvent(models.Model):
     partner_mobile = fields.Char(compute="_compute_contact")
     partner_contact_name = fields.Char(compute="_compute_contact")
     partner_city = fields.Char(related="partner_id.city")
+    # To set whatsapp URIs
```
Comment (chienandalu) [inline sale_planner_calendar/models/sale_planner_calendar_event.py]: Explain why in the compute method
```suggestion
```

File: sale_planner_calendar/models/sale_planner_calendar_event.py [hunk 3]
```diff
@@ -73,6 +78,12 @@ class SalePlannerCalendarEvent(models.Model):
...
+    sanitized_partner_mobile = fields.Char(compute="_compute_sanitized_partner_mobile")
+    # Geolocation
+    location_url = fields.Char(compute="_compute_location_url")
+    # Profile icon
```
Comment (chienandalu) [inline sale_planner_calendar/models/sale_planner_calendar_event.py]: The field name says so...
```suggestion
```

File: sale_planner_calendar/models/sale_planner_calendar_event.py [hunk 4]
```diff
@@ -73,6 +78,12 @@ class SalePlannerCalendarEvent(models.Model):
     partner_mobile = fields.Char(compute="_compute_contact")
     partner_contact_name = fields.Char(compute="_compute_contact")
     partner_city = fields.Char(related="partner_id.city")
+    # To set whatsapp URIs
+    sanitized_partner_mobile = fields.Char(compute="_compute_sanitized_partner_mobile")
+    # Geolocation
```
Comment (chienandalu) [inline sale_planner_calendar/models/sale_planner_calendar_event.py]: Explain it in the compute method
```suggestion
```

File: sale_planner_calendar/models/sale_planner_calendar_event.py [hunk 5]
```diff
@@ -130,6 +141,46 @@ def _compute_contact(self):
...
 
+    def _compute_sanitized_partner_mobile(self):
+        for rec in self:
+            rec.sanitized_partner_mobile = False
+            if rec.partner_mobile:
+                rec.sanitized_partner_mobile = re.sub(r"\W+", "", rec.partner_mobile)
+
+    def _compute_location_url(self):
+        for rec in self:
+            rec.location_url = False
```
Comment (chienandalu) [inline sale_planner_calendar/models/sale_planner_calendar_event.py]:
```suggestion
        self.location_url = False
        for rec in self:
```

File: sale_planner_calendar/models/sale_planner_calendar_event.py [hunk 6]
```diff
@@ -130,6 +141,46 @@ def _compute_contact(self):
...
+            url = ""
```
Comment (chienandalu) [inline sale_planner_calendar/models/sale_planner_calendar_event.py]: You don't need this variable, just write directly in the record field

File: sale_planner_calendar/models/sale_planner_calendar_event.py [hunk 7]
```diff
@@ -130,6 +141,46 @@ def _compute_contact(self):
...
+            event_location = rec.calendar_event_id.location
+            partner_latitude = rec.partner_id.partner_latitude
+            partner_longitude = rec.partner_id.partner_longitude
+            partner_location = rec.partner_city + rec.partner_street
+            if event_location:
+                url = event_location.replace(" ", "+")
+            elif partner_latitude != 0 or partner_longitude != 0:
+                url = partner_latitude + "%2C" + partner_longitude
```
Comment (chienandalu) [inline sale_planner_calendar/models/sale_planner_calendar_event.py]:
```suggestion
                url = f"{partner_latitude}%2C{partner_longitude}"
```

File: sale_planner_calendar/models/sale_planner_calendar_event.py [hunk 8]
```diff
@@ -130,6 +141,46 @@ def _compute_contact(self):
...
+        ctx.update(
+            {
+                "default_partner_id": self.partner_id.id,
+            }
+        )
+        action["context"] = ctx
+        domain = safe_eval(action["domain"])
+        domain.append(
+            ("partner_id", "=", self.partner_id.commercial_partner_id.id),
+        )
```
Comment (chienandalu) [inline sale_planner_calendar/models/sale_planner_calendar_event.py]: Better use `expression.AND`

**Review by chienandalu — CHANGES_REQUESTED** [review]
Some code review and light testing:

- Add a description of tasks done in the commit
- Also add in the readme the new relevant features and how to use them
- Testing I came across this error:
```
    url = partner_latitude + "%2C" + partner_longitude
Exception

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/opt/odoo/odoo/http.py", line 658, in _handle_exception
    return super(JsonRequest, self)._handle_exception(exception)
  File "/opt/odoo/odoo/http.py", line 301, in _handle_exception
    raise exception.with_traceback(None) from new_cause
TypeError: unsupported operand type(s) for +: 'float' and 'str'
```

**Review by pedrobaeza — APPROVED** [review]
You can add a before/after screenshot also for reference.

**Comment by pedrobaeza** [comment]
@pilarvargas-tecnativa please fw-port it to v16

**Pull request merged**